drv/fpga-api/src/lib.rs

```diff
@@ -267,6 +267,28 @@ impl FpgaUserDesign {
     }
 }
 
+/// Poll the device state of the FPGA to determine if it is ready to receive
+/// a bitstream, resetting the device if needed.
+pub fn await_fpga_ready(
+    fpga: &mut Fpga,
+    sleep_ticks: u64,
+) -> Result<DeviceState, FpgaError> {
+    let mut state = fpga.state()?;
+
+    while match state {
```

Comment:
Nit: this could be
```
while match!(state, DeviceState::Unknown | DeviceState::Disabled | DeviceState::Error) {
```
(using the `match!` macro and removing one level of inversion, since I found it confusing at first glance)

Reply:
Alternatively, remove the separate `mut state` variable entirely and do everything in the loop, e.g.
```rust
loop {
    let state = fpga.state()?;
    match state {
        DeviceState::AwaitingBitstream | DeviceState::RunningUserDesign => return Ok(state),
        _ => (),
    }
    fpga.reset()?;
    userlib::hl::sleep_for(sleep_ticks);
}
```
This hides the loop condition in the `match` statement, but makes everything else a little cleaner, so I don't feel too strongly about it...

Reply:
I keep forgetting there's a plain `loop {..}`. The `while` is a bit awkward and this preserves the linear nature of what I am trying to accomplish.

Reply:
Thanks, this suggestion improves this function significantly. And I fixed the edge case where the device may be held in reset, which would otherwise cause this function to hang indefinite.